scripts/automatic_tag.go

```diff
@@ -0,0 +1,246 @@
+		os.Exit(1)
+	}
+
+	updateType := os.Args[1]
+	valid := false
+	for _, t := range updateTypes {
+		if updateType == t {
+			valid = true
+			break
+		}
+	}
+	if !valid {
+		fmt.Printf("the provided argument for update type `%s` is not valid, the valid values are: [%s]\n", updateType, strings.Join(updateTypes, ","))
+		os.Exit(1)
+	}
+
+	sshPrivateKeyFilePath := os.Args[2]
+	_, err := os.Stat(sshPrivateKeyFilePath)
+	if err != nil {
+		fmt.Printf("The provided private key file path %s is not valid: %s\nUsage: %s\n", sshPrivateKeyFilePath, err, usage)
+		os.Exit(1)
+	}
+
+	sshKeyPassword := os.Args[3]
+	if sshKeyPassword == coreOnlyFlag {
+		fmt.Printf("Wrong arguments. Usage: %s\n", usage)
+		os.Exit(1)
+	}
+
+	// Parse flag
+	var coreOnly bool
+	if len(os.Args) == 5 {
+		if os.Args[4] != coreOnlyFlag {
+			fmt.Printf("Wrong arguments. Usage: %s\n", usage)
+			os.Exit(1)
+		}
+		coreOnly = true
+	}
+
+	err = automaticTagUpdate(updateType, sshPrivateKeyFilePath, sshKeyPassword, coreOnly)
+	if err != nil {
+		fmt.Printf("Error updating tags: %s\n", err.Error())
+		os.Exit(1)
+	}
+}
+
+func automaticTagUpdate(updateType, sshPrivateKeyFilePath, password string, coreOnly bool) error {
+	tempDir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return fmt.Errorf("create temporary directory: %w", err)
+	}
+
+	defer func() {
+		tempErr := os.RemoveAll(tempDir)
+		if tempErr != nil && err == nil {
+			err = fmt.Errorf("temporary directory %s could not be removed: %w", tempDir, tempErr)
+		}
+	}()
+
+	publicKeys, err := ssh.NewPublicKeysFromFile("git", sshPrivateKeyFilePath, password)
+	if err != nil {
+		return fmt.Errorf("get public keys from private key file: %w", err)
+	}
+
+	r, err := git.PlainClone(tempDir, false, &git.CloneOptions{
+		Auth:              publicKeys,
+		URL:               sdkRepo,
+		RecurseSubmodules: git.DefaultSubmoduleRecursionDepth,
+	})
+	if err != nil {
+		return fmt.Errorf("clone SDK repo: %w", err)
+	}
+
+	tagrefs, err := r.Tags()
+	if err != nil {
+		return fmt.Errorf("get tags: %w", err)
+	}
+
+	latestTags := map[string]string{}
+	err = tagrefs.ForEach(func(t *plumbing.Reference) error {
+		latestTags = storeLatestTag(t, latestTags, coreOnly)
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("iterate over existing tags: %w", err)
+	}
+
+	for service, version := range latestTags {
+		newTag, err := computeUpdatedTag(service, version, updateType, coreOnly)
+		if err != nil {
+			fmt.Printf("Error for %s with version %s, this tag will be skipped. Error: %s\n", service, version, err.Error())
+			continue
+		}
+
+		err = createTag(r, newTag)
+		if err != nil {
+			fmt.Printf("Create tag %s returned error: %s\n", newTag, err)
+			continue
+		}
+		fmt.Printf("Created tag %s\n", newTag)
+	}
+
+	err = pushTags(r, publicKeys)
+	if err != nil {
+		return fmt.Errorf("push tags: %w", err)
+	}
+	return nil
+}
+
+func storeLatestTag(t *plumbing.Reference, latestTags map[string]string, coreOnly bool) map[string]string {
```

Comment:
I think would be good to have some short descriptions of what these functions do and what is the use of each argument

Reply:
As it might not be obvious just from the name and also because some types of the `go-git` package (e.g. `plumbing.Reference`) are a bit obscure